### PR TITLE
feat: Add support for config files in ES module format

### LINKED ISSUE #2604: web-ext-config ES Module Support
### Is this a feature request or a bug?

Feature

### What is the current behavior?

If you've migrated your browser extension to no longer use CommonJS modules then Node will no longer parse the `web-ext-config.js` file without additional changes.

This is the correct behavior as per the `web-ext` docs:

>The file should be a CommonJS module [as understood by NodeJS](https://nodejs.org/docs/latest/api/modules.html#modules_modules) and must export each configuration value.

and

>Automatic discovery of configuration files
>
>web-ext will load existing configuration files in the following order:
>
> * A config file named `.web-ext-config.js` in your home directory, for example: - On Windows: `C:\Users\<username>\.web-ext-config.js` - On macOS: `/Users/<username>/.web-ext-config.js` - On Linux: `/home/<username>/.web-ext-config.js`
> * A config property named `"webExt"` included in a `package.json` file in the current directory.
> * A config file named `web-ext-config.js` in the current directory.

https://extensionworkshop.com/documentation/develop/getting-started-with-web-ext/#setting-option-defaults-in-a-configuration-file

### Example

In `package.json` you have set:

```json
{
  "type": "module",
}
```

and you then run any `web-ext` command you get an error from Node.

```$shell
$ npx web-ext lint
Applying config files: ./package.json, ./web-ext-config.js

UsageError: Cannot read config file: /REDACTED/web-ext-config.js
Error: require() of ES Module /REDACTED/web-ext-config.js from /REDACTED/node_modules/import-fresh/index.js not supported.
web-ext-config.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename web-ext-config.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /REDACTED/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).
```

### What is the expected or desired behavior?

Quickest win would be to add `./web-ext-config.cjs` to the pool of default files checked.

Less quick would be to allow ES Modules `export {}` in `web-ext-config.{,m}js`.

The workaround for now is to add `--config web-ext-config.cjs` as a suffix to `web-ext` commands.

It seems setting `config` in the `webExt` block in `package.json` _doesn't_ work:

```json
{
  "webExt": {
    "config": "web-ext-config.cjs"
  }
}
```

### Version information (for bug reports)

```
node --version && npm --version && npx web-ext --version
v16.15.1
8.11.0
7.4.0
```

### PR DESCRIPTION
**Pull request by rpl**
This PR includes a small set of changes to:
- add support for loading config files in the ESM module format
- add guidance help message to UserError error messages raised when hitting parsing error because the config file is not in the nodejs instance expects it to be (e.g. either a .js config file in CommonJS format from a type module package, or a .js config file in ES module format from a type commonjs package)
- proposed addition of an additional command which does just print all config data loaded from the discovered files

Fixes #2604 

TODO:
- [x] agree with the team if we would prefer the behavior on .js config files to be just throwing and force the user to conver them to either .cjs or .mjs config files (instead of allow them to load successfully whenever possible)
- [x] agree with the team if we feel that introducing an additional command which just dump the config data as discovered and loaded (mainly to make it easier for extensions developers to inspect the resulting data got from loading and merging all discovered config files, e.g. to debug and/or report to us issue with config file loading)
  - [x] either cover the new command with a few new tests or remove the command prototype all together (based on which is the agreement we reached on the parent todo item)

### CODE REVIEW

**Comment by rpl** [comment]
@willdurand this PR includes an initial draft of a proposed set of changes to handle the issue tracked by #2604 

As briefly mentioned in the task list at the end of https://github.com/mozilla/web-ext/pull/3029#issue-2094469693

- the current version of this PR still allows and loads config data from .js config files, if we are able to load them successfully, and show an help message to guide the developers with config files that would not parse for mismatching between the format of the config file (ESM or CommonJS) and the one that nodejs expects. Alternatively we could consider to still discover .js config file but forcefully fail to force their conversion into explicit .mjs or .cjs config files. Let me know your thoughts around this two possible behaviors (or if you had some other behavior that may be worth considering)

- the new command is not stricly necessary but I was considering the idea while drafting this PR and so I thought to include it in a separate commit to bring it up and touch base about it (and so that I can just remove the commit if we agree that we are not going to add it).

**Comment by rpl** [comment]
@willdurand this is ready for another pass from your perspective, this updates includes the changes we agreed on yesterday and few more small things:
- added a warning on loading from .js config file
- renamed and tweaked the dump command to emit only valid json to stdout as we discussed yesterday and covered it with a basic test case
- removed the import-fresh dependency while I was here.

File: src/config.js [hunk 1]
```diff
@@ -1,16 +1,38 @@
...
+
+export const HELP_ERR_MODULE_FROM_ESM = [
+  'This config file belongs to a package.json file with "type" set to "module".',
+  'Rename the config file to end in .cjs or rename it to .mjs and rewrite it to an ES module.',
+].join(' ');
+
+export const HELP_ERR_IMPORTEXPORT_CJS = [
+  'This config file is defined as an ESM module, but it belongs to either a project dir',
+  'with a package.json file with "type" set to "commonjs" or one without any package.json file.',
+  'Rename the config file to end in ".mjs" to fix the config loading error.',
```
Comment (willdurand) [inline src/config.js:29]:
```suggestion
  'Change the file extension to ".mjs" to fix the config loading error.',
```

File: src/config.js [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
 
 import fileExists from './util/file-exists.js';
 import { createLogger } from './util/logger.js';
 import { UsageError, WebExtError } from './errors.js';
 
 const log = createLogger(import.meta.url);
 
+export const WARN_LEGACY_JS_EXT = [
+  'should be renamed to .cjs or .mjs to ensure its format is not ambiguous.',
+  'Config files with .js extensions are deprecated and will not be loaded anymore',
```
Comment (willdurand) [inline src/config.js:17]:
```suggestion
  'Config files with the ".js" file extension are deprecated and will not be loaded anymore',
```

File: src/config.js [hunk 3]
```diff
@@ -1,16 +1,38 @@
...
+import parseJSON from 'parse-json';
 
 import fileExists from './util/file-exists.js';
 import { createLogger } from './util/logger.js';
 import { UsageError, WebExtError } from './errors.js';
 
 const log = createLogger(import.meta.url);
 
+export const WARN_LEGACY_JS_EXT = [
+  'should be renamed to .cjs or .mjs to ensure its format is not ambiguous.',
```
Comment (willdurand) [inline src/config.js:16]:
```suggestion
  'should be renamed to use the ".cjs" or ".mjs" file extension to ensure its format is not ambiguous.',
```

Comment (willdurand) [inline src/config.js:16]: Maybe add a comment to indicate that this message is added to another string when it is printed. Other strings below are full sentences but this one is partial.

File: src/config.js [hunk 4]
```diff
@@ -115,20 +137,49 @@ export function applyConfigToArgv({
...
+      configModule = parseJSON(
+        await fs.readFile(resolvedFilePath, { encoding: 'utf-8' }),
+      );
+    } else {
+      configModule = await import(`file://${resolvedFilePath}?nonce=${nonce}`);
+    }
+
+    if (configModule.default) {
+      const { default: configDefault, ...esmConfigMod } = configModule;
+      // ESM modules may expose both a default and named exports and so
```
Comment (willdurand) [inline src/config.js:163]:
```suggestion
      // ES modules may expose both a default and named exports so
```

File: src/config.js [hunk 5]
```diff
@@ -115,20 +137,49 @@ export function applyConfigToArgv({
...
+      // we merge the named exports of top of what may have been set in
```
Comment (willdurand) [inline src/config.js:164]:
```suggestion
      // we merge the named exports on top of what may have been set in
```

File: src/program.js [hunk 6]
```diff
@@ -521,6 +521,12 @@ Example: $0 --help run.
         },
       },
     )
+    .command(
+      'dump-config',
+      'Run config discovery and dump the resulting config data',
```
Comment (willdurand) [inline src/program.js:526]:
```suggestion
      'Run config discovery and dump the resulting config data as JSON',
```

(or something like that)

File: src/util/logger.js [hunk 7]
```diff
@@ -31,6 +31,8 @@ export class ConsoleStream {
       const msg = this.format(packet);
       if (this.isCapturing) {
         this.capturedMessages.push(msg);
+      } else if (packet.level > bunyan.INFO) {
+        localProcess.stderr.write(msg);
```
Comment (willdurand) [inline src/util/logger.js:35]: ❤️

File: src/program.js [hunk 8]
```diff
@@ -521,6 +521,12 @@ Example: $0 --help run.
...
+      commands.dumpConfig,
+      {},
+    )
```
Comment (willdurand) [inline src/program.js:529]: I filed an issue to document this command eventually: https://github.com/mozilla/extension-workshop/issues/1784

**Review by willdurand — APPROVED** [review]
r+wc, thanks

File: src/cmd/dump-config.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { createLogger } from '../util/logger.js';
+
+const log = createLogger(import.meta.url);
+
+export default async function config(configData) {
+  log.info(`${JSON.stringify(configData, null, 2)}`);
```
[the remark this answers is not in the document]
Comment (rpl) [reply src/cmd/dump-config.js:6]: yep, that is definitely a left over :facepalm:

File: src/program.js (hunk 8 above)
Comment (rpl) [reply src/program.js:529]: Thank you for taking care of filing that followup as in issue in the extension-workshop! :heart:

**Pull request merged**